Add modular exponentiation(3-argument pow) for Expr

ylemkimon opened the pull request:
Added modular exponentiation(3-argument pow) support for Expr.

Fixes #5715.

ylemkimon committed:
Add modular exponentiation(3-argument pow) for Expr

Added modular exponentiation(3-argument pow) support for Expr.

Upabjojr commented:
I don't think this is the right way to implement this feature.

Upabjojr commented:
This code should be inside Mod. You would probably want to call `Mod(a**b, m)`

ylemkimon commented:
@Upabjojr Mod does that for unevaluated `Pow`. It was added in #13236.

Upabjojr commented:
So maybe we need a new class, say `PowMod` or `ModPow`. Or a new method in `Pow`. I wouldn't hack `__pow__` to achieve this.

ylemkimon commented:
@Upabjojr Python has built-in [3-argument pow(modular exponentiation)](https://docs.python.org/3/library/functions.html#pow).

Upabjojr commented:
Oh, I see.

Upabjojr commented on sympy/core/expr.py:
```diff
@@ -144,9 +144,25 @@ def __mul__(self, other):
     @_sympifyit('other', NotImplemented)
     @call_highest_priority('__rpow__')
-    def __pow__(self, other):
+    def ___pow__(self, other):
```
Why did you rename this?

ylemkimon replied:
@Upabjojr New `__pow__` calls `___pow__`, the original `__pow__`, if mod is None or not all arguments are integers. (See line 149 and 158 above)

Upabjojr replied:
I've seen this. But why not just keep one function?

ylemkimon replied:
@Upabjojr It’s because _sympifyit wrapper and call_highest_priority doesn’t support 3-argument function.

jksuom replied:
Perhaps this could be something like `_pow` as it is a private name. (Three underscores feels excessive:)

ylemkimon replied:
@jksuom Renamed ___pow__ to _pow and reordered functions.

ylemkimon committed:
Rename ___pow__ to _pow and reorder functions

Renamed ___pow__ to _pow and reordered functions.

jksuom commented on sympy/core/expr.py:
```diff
@@ -146,9 +146,25 @@ def __rmul__(self, other):
+                return mod_inverse(pow(_self, -other, mod), mod)
+        except ValueError:
+            power = self._pow(other)
+            if power is NotImplemented:
```
Can this really be true?

ylemkimon replied:
@jksuom `_pow` returns `NotImplemented` when _sympifying `other` fails, e.g., `pow(S(2), "2", 3)`. If we don't catch it, as `%`(mod) operation is not defined for `NotImplementedType`, it raises `TypeError: unsupported operand type(s) for %: 'NotImplementedType' and 'type(mod)'`, which I think is confusing and less informative. If we catch it and return `NotImplemented`, python raises `TypeError: unsupported operand type(s) for pow(): 'Integer', 'str', 'int'`.

smichr replied:
> it raises TypeError: unsupported operand type(s)

Then should the code be
```python
try:
    power = self._pow(other)
except TypeError:
    raise NotImplementedError  <--note NotImplementedError instead of NotImplemented?
```

And a test like `raises(NotImplementedError, lambda: pow(...failing args...))` should be added.

ylemkimon replied:
@smichr `_pow` itself does not raises TypeError, rather it returns `NotImplemented` when _sympify fails. Python automatically constructs and raises `TypeError` if `NotImplemented` is returned by overridden default operations, e.g., `__pow__`.

jksuom replied:
> _pow returns NotImplemented when _sympifying other fails, 

I didn't know that. I thought that `SympifyError` or `NotImplementedError` was raised if `other` is not sympifiable in `Pow(self, other)`.

smichr replied:
Can we get a test for this added to the suite so this line is covered?

ylemkimon replied:
@smichr Added 😄

ylemkimon replied:
Changed to catch TypeError, for when `mod` is not _sympifiable.

asmeurer commented on sympy/core/tests/test_arit.py:
```diff
@@ -187,6 +187,13 @@ def test_pow3():
+def test_mod_pow():
+    assert pow(S(4), 13, 497) == pow(4, 13, 497)
+    assert pow(S(4), -3, 497)*4**3%497 == 1
+    assert pow(x, y, z) == x**y%z
```
Can you test all combinations of SymPy/Python input types? I don't remember exactly how Python does dispatch for three-argument `__pow__`, but we should make sure all the combinations are tested.

asmeurer replied:
You might also add a test for something like `pow(S(2), S(10000000000), S(3))` (you can add more zeros if necessary), to make sure that it is using the efficient algorithm.

ylemkimon replied:
@asmeurer I'm not sure how and whether to test that it is using efficient algorithm, since it depends on Python implementation.

asmeurer replied:
`pow(S(2), S(10000000000), S(3))` will do the trick. If you try to evaluate `2**10000000000 % 3` naively it won't work. Using the Python implementation is fine. This just tests that it does indeed do it (or something equivalent).

ylemkimon replied:
@asmeurer Added tests for various types 😃

ylemkimon committed:
Change to catch TypeError in _pow

Changed to catch TypeError during power%mod in _pow for when mod is not
sympifiable.

ylemkimon committed:
Add/Change test cases for test_mod_pow

Added/Changed test cases for test_mod_pow.

ylemkimon committed:
Add test for modular exponentiation of large number

Added test for modular exponentiation of very large number to test that
it uses built-in modular exponentiation.

ylemkimon commented:
@asmeurer I missed your comment, sorry. Added test for very large numbers. But may there be a Python implementation other than Cython that does not implement efficient modular exponentiation?

smichr commented:
@ylemkimon , does your question to Aaron mean that you want to hold off on committing this?

ylemkimon commented:
@smichr No, it's good to merge. Thank you.

asmeurer commented:
> But may there be a Python implementation other than CPython that does not implement efficient modular exponentiation?

I kind of doubt that. The whole point of having a three-argument pow() is to be able to compute modular exponents fast. The naive method could take forever to compute, or even crash the interpreter by running out of memory if `a**b` is too large. If a Python implementation implements three-argument pow() inefficiently a lot more code than just this test will fail under it.

asmeurer commented:
Thanks. It's awesome to see this finally implemented.
